designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,284 @@
+3. Create an options object that is a shallow clone of the first non-`undefined` from `ruleConfig` or `schema.default`, or `{}` if the schema contains `type: "object"`
+4. If the options is truthy and schema is an object:
+   1. Recursively apply defaulting logic for each of the schema's `properties`
+
+Draft PR: [[Reference] feat!: factor in schema defaults for rule options](https://github.com/eslint/eslint/pull/17580)
+
+### Out of Scope
+
+#### Default `anyOf` Values
+
+JSON schemas can use `anyOf` to indicate values whose types are analogous to TypeScript union types.
+Resolving which `anyOf` constituent a runtime config aligns to can be complex.
+It is out of scope for this RFC to investigate heuristics for that problem.
+
+#### TypeScript Types
+
+Right now, there are two community strategies for declaring rule options types, both with drawbacks:
+
+- [`@types/eslint`](https://npmjs.com/package/@types/eslint) (the community-authored package describing ESLint's types for TypeScript users) [declares `RuleContext`'s `options` as `any[]`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ca231578e79ab1c2f01a308615e60a432061c97a/types/eslint/index.d.ts#L703), indicating it can be an array of any value types.
+  - Drawback: `any` is generally considered a bad practice for TypeScript users as it provides no type descriptions or safety checks.
+- [`@typescript-eslint/utils` > `RuleCreator`](https://typescript-eslint.io/developers/custom-rules#rulecreator) (the recommended TypeScript-enabled rule development utility) declares a [generic `RuleContext<TMessageIds, TOptions>` type with an `options: TOptions` property](https://github.com/typescript-eslint/typescript-eslint/blob/72149e1fe8f57fd217dfc59295a51df68187aad4/packages/utils/src/ts-eslint/Rule.ts#L175).
+  - Drawback: `RuleContext` already doesn't factor in default options to the type system (https://github.com/typescript-eslint/typescript-eslint/issues/5439)
+
+Resolving the type system type resulting from a fixed JSON schema -with or without `default`- and a user-provided object can be complex.
+It is out of scope for this RFC to investigate heuristics for that problem.
+
+## Documentation
+
+We'll want to augment the following docs generators in:
+
+- The core ESLint docs: automating and therefore standardizing how rule docs pages phrase default values
+- typescript-eslint's docs: the same as the core ESLint docs
+- [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator)
+
+We'll also want to mention this in the ESLint core custom rule documentation.
+
+## Drawbacks
+
+- This increases both conceptual and implementation complexity around rule options
+- As a breaking change, this may break user rules - especially private rules we don't have visibility to
+
+## Backwards Compatibility Analysis
```

Comment:
Instead of adding `context.optionsRaw`, what if we allowed plugins to export a property named `useRuleOptionDefaults` (or similar name)? This RFC's functionality would be hidden behind this option (default false during ESLint v8) so that it wouldn't initially be a breaking change and so that we could get it into the hands of plugin authors for testing and usage sooner. We could then enable the option by default as a minimal-code breaking change in the next major release (and ideally remove the option at that time or in a future major release so that we wouldn't have to support both modes forever).

Reply:
I like this a lot! Adding in.

Though, it wouldn't replace the need for `optionsRaw`, as the three noted use cases for it are all in ESLint core. I'd assume we'd want core to use both the new recursively default `context.options` and the traditional `context.optionsRaw`.